src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+                            maskingKeySpan = maskingKey.First.Span;
+                        }
+                        else
+                        {
+                            // We need to copy it to a single span.
+                            maskingKeySpan = new Span<byte>(maskingKeyStorage, 4);
+                            maskingKey.CopyTo(maskingKeySpan);
+                        }
+
+                        // Either we initialized it with a length, or it came from the maskingKey buffer which should already have been sliced.
+                        Debug.Assert(maskingKeySpan.Length == 4);
+
+                        // Unmask
+                        for (int i = 0; i < payload.Length; i++)
+                        {
+                            unmasked[i] ^= maskingKeySpan[i % 4];
+                        }
+                        payload = ReadableBuffer.Create(unmasked, 0, unmasked.Length);
+                    }
+
+                    // Mark the masking key consumed
+                    _inbound.Advance(buffer.Start);
+                }
+            }
+            else
+            {
+                // If there was a masking key, mark it consumed
+                if (masked)
+                {
+                    _inbound.Advance(buffer.Start);
+                }
+
+                // Clear the "lastFrameCursor" so we don't try to advance anything next receive
+                _lastPayloadEnd = null;
+            }
+
+            return new WebSocketFrame(fin, opcode, payload);
+        }
+
+        /// <summary>
+        /// Sends the specified frame.
+        /// </summary>
+        /// <param name="message">The message to send.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> that indicates when/if the send is cancelled.</param>
+        /// <returns>A <see cref="Task"/> that completes when the message has been written to the outbound stream.</returns>
+        // TODO: De-taskify this to allow consumers to create their own awaiter.
+        public Task SendAsync(WebSocketFrame message, CancellationToken cancellationToken)
+        {
+            // Calculate the size of the frame
+            // Base header size is 2 bytes.
+            var allocSize = 2;
+            var payloadLength = message.Payload.Length;
+            if (payloadLength > ushort.MaxValue)
+            {
+                // We're going to need an 8-byte length
+                allocSize += 8;
+            }
+            else if (payloadLength > 125)
+            {
+                // We're going to need a 2-byte length
+                allocSize += 2;
+            }
+            if (_maskingKey != null)
+            {
+                // We need to alloc for the whole payload (and the masking key)
+                allocSize += 4 + payloadLength;
+            }
+
+            // Allocate a buffer
+            var buffer = _outbound.Alloc(minimumSize: allocSize);
+            if (buffer.Memory.Length < allocSize)
```

Comment:
If we do masking in-place we don't need a huge alloc. Originally this code only alloc'd enough for the header, it was only when masking was added (and it ended up requiring a copy, because we don't have a clear ownership semantic on the ReadableBuffer right now) that it ballooned out like this :)